### PR TITLE
AP_Arming: allow bypassing compass healthy check when external navigation data is available for heading

### PR DESCRIPTION
**Pull request by hoangthien94**
**Problem**: an external source of heading (SLAM, mocap, vision etc.) can be used as yaw in place of the compass. As such, when compass is disabled and external navigation data is available, the arming check for compass health can be skipped completely. This PR provides a way to do this.

**Approach**:
`extNavUsedForYaw`, defined here:
https://github.com/ArduPilot/ardupilot/blob/bab31a2d61828e034a5e42440a0604f86b6b998b/libraries/AP_NavEKF2/AP_NavEKF2_core.h#L1124 seems to be the perfect candidate for the job. This variable will be set to `true` whenever the EKF core sees that external navigation data is available *and* compass is disabled.
https://github.com/ArduPilot/ardupilot/blob/634db441f885c8bff57bd98c02c2ce7c2896c1d8/libraries/AP_NavEKF2/AP_NavEKF2_PosVelFusion.cpp#L317
Thus, we will check this variable of the primary EKF core. If `true`, we skip the compass health check completely. To facilitate this, we add functions to obtain the variable's value.

**Related**: issue and forum discussion:
- https://github.com/ArduPilot/ardupilot/issues/5454
- https://discuss.ardupilot.org/t/compass-not-healthy-preventing-arming-even-though-compasses-are-disabled-using-slam-instead/45011/9

### CODE REVIEW

**Comment by rmackay9** [comment]
@hoangthien94, thanks very much for this!

I wonder if we should perhaps expose this to the AHRS object and have DCM and EKF3 always return false.  The reason is that I don't think we know if we are using EKF2 or EKF3.  So if the AHRS_EKF_TYPE = 3 but we had EKF2 also running (and receiving external position estimates) then it would return true and we would skip over the compass checks when in fact, EKF3 needs a compass to work.  This is an extremely unlikely case but I think it shows the accessor should be one level higher..

**Comment by hoangthien94** [comment]
The check is now one level higher. Return value depends on current `ekf_type()`, which is `false` for now in the case of both DCM and EKF3. 
I guess one support for external navigation goes into EKF3, we will have to come back and revise this.

**Comment by rmackay9** [comment]
much better, thanks!  I wonder if we should name the accessor, "external_nav_used_for_yaw"?

**Comment by hoangthien94** [comment]
I agree that the function name should be a bit more descriptive of its action and follow the same style as others in the same library. I will make a change.

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp [hunk 1]
```diff
@@ -1821,5 +1821,22 @@ AP_AHRS_NavEKF &AP::ahrs_navekf()
...
+// check whether compass can be bypassed for arming check in case when external navigation data is available 
+bool AP_AHRS_NavEKF::is_compass_bypassable(void) const
+{
+    // for DCM and EKF3, always return false for now.
+    switch (ekf_type()) {
+    case EKF_TYPE_NONE:
+        return false;
+
+    case EKF_TYPE2:
+    default:
```
[the remark this answers is not in the document]
Comment (hoangthien94) [reply libraries/AP_AHRS/AP_AHRS_NavEKF.cpp]: I forgot about this one. I have removed the other EKF types, now it's just EKF2 and default (always return false).

**Comment by OXINARF** [comment]
@hoangthien94 You squashed a bit too much 🙂 I've fixed it for you and pushed it here. I'll try to get @priseborough's opinion on whether the EKF2 method should be in the Outputs.cpp file.

**Comment by hoangthien94** [comment]
I see, so each commit should be about one library, whereas I just squashed everything into one. Thank you @OXINARF for your help.

**Comment by OXINARF** [comment]
Yes, that's right. Two other tips:

- pay attention to commit message length, preferably it shouldn't be bigger than 72 chars so that it displays in GitHub in one line; sometimes it's not easy (for example, I shortened the last one here, but it still passed the limit). Note that commit sub-messages (any text you write after a blank line) can have whatever size you want
- when squashing: the commit message should be looked at again because sometimes it's not accurate anymore or includes unnecessary sub-messages

Feel free to ask questions when in doubt, either in the PR or in our Gitter chat (http://gitter.im/ArduPilot/ardupilot).

File: libraries/AP_AHRS/AP_AHRS_NavEKF.cpp [hunk 2]
```diff
@@ -1821,5 +1821,17 @@ AP_AHRS_NavEKF &AP::ahrs_navekf()
     return static_cast<AP_AHRS_NavEKF&>(*AP_AHRS::get_singleton());
 }
 
+// check whether compass can be bypassed for arming check in case when external navigation data is available 
+bool AP_AHRS_NavEKF::is_ext_nav_used_for_yaw(void) const
```
Comment (priseborough) [inline libraries/AP_AHRS/AP_AHRS_NavEKF.cpp]: Can we please move this function to Outputs.cpp

File: libraries/AP_NavEKF2/AP_NavEKF2_core.cpp [hunk 3]
```diff
@@ -1616,3 +1616,8 @@ void NavEKF2_core::zeroAttCovOnly()
     }
 }
 
+// return true when external nav data is also being used as a yaw observation
```
Comment (priseborough) [inline libraries/AP_NavEKF2/AP_NavEKF2_core.cpp]: Can we please move this to Outputs.cpp

**Comment by OXINARF** [comment]
@rmackay9 are you OK with merging?

**Comment by rmackay9** [comment]
@OXINARF, yes, looks good to me, thanks!

**Comment by xjka** [comment]
When will the new binaries be available with this patch? (For Arducopter for e.g.?)

**Comment by juri117** [comment]
What about EKF3? I would like to provide yaw using the mavlink GPS_INPUT message and disable the compass. However, that only works with EKF3 (EK3_MAG_CAL=5). Arming fails with "Compass not healthy" because libraries/AP_AHRS/AP_AHRS_NavEKF.cpp -> is_ext_nav_used_for_yaw() returns const false for EKF3.

**Comment by peterbarker** [comment]
@juri117 good news - EKF3 can now do what you want.

**Pull request merged**